R/install_torsten.R

```diff
@@ -0,0 +1,197 @@
+#' Install Torsten
+#'
+#' @description The `install_torsten()` function attempts to download and
+#'   install the latest release of [Torsten](https://github.com/metrumresearchgroup/Torsten/releases/latest).
+#'   Installing a previous release or a new release candidate is also possible
+#'   by specifying the `version` or `release_url` argument.
+#'   See the first few sections of the CmdStan
+#'   [installation guide](https://mc-stan.org/docs/cmdstan-guide/cmdstan-installation.html)
+#'   for details on the C++ toolchain required for installing CmdStan.
+#'
+#'
+#' @export
+#' @param dir (string) The path to the directory in which to install Torsten
+#'   The default is to install it in a directory called `.torsten` within the
+#'   user's home directory (i.e, `file.path(Sys.getenv("HOME"), ".torsten")`).
+#' @param cores (integer) The number of CPU cores to use to parallelize building
+#'   CmdStan and speed up installation. If `cores` is not specified then the
+#'   default is to look for the option `"mc.cores"`, which can be set for an
+#'   entire \R session by `options(mc.cores=value)`. If the `"mc.cores"` option
+#'   has not been set then the default is `2`.
+#' @param quiet (logical) Should the verbose output
+#'   from the system processes be suppressed when building the Torsten binaries?
+#'   The default is `FALSE`.
+#' @param version (string) The Torsten release version to install. The default
+#'   is `NULL`, which downloads the latest stable release from
+#'   <https://github.com/metrumresearchgroup/Torsten/releases>.
+#' @param release_url (string) The URL for the specific Torsten release or
+#'   release candidate to install. See <https://github.com/metrumresearchgroup/Torsten/releases>.
+#'   The URL should point to the tarball (`.tar.gz.` file) itself, e.g.,
+#'   `release_url="https://github.com/metrumresearchgroup/Torsten/archive/refs/tags/torsten_v0.90.0.tar.gz"`.
+#'   If both `version` and `release_url` are specified then `version` will be used.
+#' @param cpp_options (list) Any makefile flags/variables to be written to
+#'   the `make/local` file. For example, `list("CXX" = "clang++")` will force
+#'   the use of clang for compilation.
+#' @param check_toolchain (logical) Should `install_torsten()` attempt to check
+#'   that the required toolchain is installed and properly configured. The
+#'   default is `TRUE`.
+#'
+#' @examples
+#' \dontrun{
+#' # install_torsten(cores = 4)
+#'
+#' cpp_options <- list(
+#'   "CXX" = "clang++",
+#'   "CXXFLAGS+= -march=native",
+#'   PRECOMPILED_HEADERS = TRUE
+#' )
+#' # install_torsten(cores = 4, cpp_options = cpp_options)
+#' }
+#'
+
+#'## install_torsten version without unexported dependencies
+
+latest_torsten_release <- function() {
+  require(jsonlite)
+  dest_file <- tempfile(pattern = "releases-", fileext = ".json")
```

Comment:
Ideally `dest_file` would be removed after its no longer needed. Could use [withr::local\_tempfile](https://withr.r-lib.org/reference/with_tempfile.html) helper here.

Reply:
Do you mean that I should replace `tempfile` with `withr::local_tempfile`? If so, I did that. The doc for `withr::local_tempfile` is pretty terse. Am I right in thinking that it will delete `dest_file` immediately upon exiting the `latest_torsten_release` function?

Reply:
> Am I right in thinking that it will delete `dest_file` immediately upon exiting the `latest_torsten_release` function?

Correct.  It's functionally equivalent to

```
f <- tempfile(...)
on.exit(unlink(f), add = TRUE)
```

Reply:
Done